src/pstore/pstore.c

```diff
@@ -0,0 +1,513 @@
+        int r = 0;
+
+        ifd_path = path_join(arg_sourcedir, pe->direntbuf.d_name);
+        if (!ifd_path)
+                return log_oom();
+
+        if (arg_storage == PSTORE_STORAGE_EXTERNAL) {
+                /* This code copies the file from the pstore to other storage.
+                 * The rename() syscall is not utilized as it results in the
+                 * Invalid cross-device link error.
+                 * In addition, an optional subdirectory can be specified in
+                 * forming the final destination path.  */
+                _cleanup_free_ char *ofd_path;
+                ofd_path = path_join(arg_archivedir, subdir, pe->direntbuf.d_name);
+                if (!ofd_path)
+                        return log_oom();
+
+                /* Make sure destination exists */
+                r = mkdir_parents(ofd_path, 0755);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to create directoy %s: %m", ofd_path);
+                r = copy_file_atomic(ifd_path, ofd_path, 0644, 0, 0, COPY_REPLACE);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to copy_file_atomic: %s to %s", ifd_path, ofd_path);
+        }
+        else
+        if (arg_storage == PSTORE_STORAGE_JOURNAL) {
+                //sd_journal_send("MESSAGE=%s", pe->direntbuf.d_name, NULL);
+                if (pe->is_binary) {
+                        //WITH_BINARY= but how to know end of binary blob if contained in 0 terminated string?
+                }
+                else {
+                        /* NOTE: If journald not running, r always 0 */
+                        r = sd_journal_send("MESSAGE=File %s:\n%s", pe->direntbuf.d_name, pe->contents, NULL);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to sd_journal_send: %s", ifd_path);
+                }
+        }
+
+        /* If file copied properly, remove it from pstore */
+        if (arg_unlink)
+                (void) unlink(ifd_path);
+
+        return 0;
+}
+
+static int write_dmesg(const char *dmesg, size_t size, const char *id) {
+        _cleanup_free_ char *ofd_path;
+        _cleanup_close_ int ofd = -1;
+        ssize_t wr;
+
+        if (isempty(dmesg) || (size == 0))
+                return 0;
+
+        log_info("Record ID %s", id);
+
+        ofd_path = path_join(arg_archivedir, id, "dmesg.txt");
+        if (!ofd_path)
+                return log_oom();
+
+        ofd = open(ofd_path, O_CLOEXEC|O_CREAT|O_TRUNC|O_WRONLY, 0644);
+        if (ofd < 0)
+                return log_error_errno(errno, "Failed to open %s: %m", ofd_path);
+
+        wr = write(ofd, dmesg, size);
+
+        if (wr < 0)
+                return log_error_errno(errno, "Failed to store dmesg to %s: %m", ofd_path);
+        if (wr != (ssize_t)size)
+                return log_error("Failed to store dmesg to %s. %zu bytes are lost.", ofd_path, size - wr);
```

Comment:
please write these files atomically, i.e. create under temporary name, write in full, then rename into place. open_tmpfile_linkable() + link_tmpfile() will help you with that and automatically use O_TMPFILE if that's possible. This has the benefit that we either we succeed in writing the files or don't write anything at all, but never leave half-written files.

Reply:
done